src/providers/openrouter/chatComplete.ts

```diff
@@ -0,0 +1,148 @@
+import { OPENROUTER } from '../../globals';
+import {
+  ChatCompletionResponse,
+  ErrorResponse,
+  ProviderConfig,
+} from '../types';
+import {
+  generateErrorResponse,
+  generateInvalidProviderResponseError,
+} from '../utils';
+
+export const OpenrouterChatCompleteConfig: ProviderConfig = {
+  model: {
+    param: 'model',
+    required: true,
+    default: 'glm-3-turbo',
+  },
+  messages: {
+    param: 'messages',
+    default: '',
+  },
+  max_tokens: {
+    param: 'max_tokens',
+    default: 100,
+    min: 0,
+  },
+  temperature: {
+    param: 'temperature',
+    default: 1,
+    min: 0,
+    max: 2,
+  },
+  top_p: {
+    param: 'top_p',
+    default: 1,
+    min: 0,
+    max: 1,
+  },
+  stream: {
+    param: 'stream',
+    default: false,
+  },
+};
+
+interface OpenrouterChatCompleteResponse extends ChatCompletionResponse {
+  id: string;
+  object: string;
+  created: number;
+  model: string;
+  usage: {
+    prompt_tokens: number;
+    completion_tokens: number;
+    total_tokens: number;
+  };
+}
+
+export interface OpenrouterErrorResponse {
+  object: string;
+  message: string;
+  type: string;
+  param: string | null;
+  code: string;
+}
+
+interface OpenrouterStreamChunk {
+  id: string;
+  object: string;
+  created: number;
+  model: string;
+  choices: {
+    delta: {
+      role?: string | null;
+      content?: string;
+    };
+    index: number;
+    finish_reason: string | null;
+  }[];
+}
+
+export const OpenrouterChatCompleteResponseTransform: (
+  response: OpenrouterChatCompleteResponse | OpenrouterErrorResponse,
+  responseStatus: number
+) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+  if ('message' in response && responseStatus !== 200) {
+    return generateErrorResponse(
+      {
+        message: response.message,
+        type: response.type,
+        param: response.param,
+        code: response.code,
+      },
+      OPENROUTER
+    );
+  }
+
+  if ('choices' in response) {
+    return {
+      id: response.id,
+      object: response.object,
+      created: response.created,
+      model: response.model,
+      provider: OPENROUTER,
+      choices: response.choices.map((c) => ({
+        index: c.index,
+        message: {
+          role: c.message.role,
+          content: c.message.content,
+        },
+        finish_reason: c.finish_reason,
+      })),
+      usage: {
+        prompt_tokens: response.usage?.prompt_tokens,
+        completion_tokens: response.usage?.completion_tokens,
+        total_tokens: response.usage?.total_tokens,
+      },
+    };
+  }
+
+  return generateInvalidProviderResponseError(response, OPENROUTER);
+};
+
+export const OpenrouterChatCompleteStreamChunkTransform: (
+  response: string
+) => string = (responseChunk) => {
+  let chunk = responseChunk.trim();
+  chunk = chunk.replace(/^data: /, '');
+  chunk = chunk.trim();
+  if (chunk === '[DONE]') {
```

Comment:
openrouter also sends a `:OPENROUTER PROCESSING` event as a part of stream. Can you please add a check just like `[DONE] ` but return `undefined` in that case. Example:
```
if (chunk === ': OPENROUTER PROCESSING') {
    return undefined;
  }
```

Reply:
![Uploading Screenshot 2024-04-10 at 18.14.56.png…]()

Reply:
you are right, I think we just return the same value. Indeed, it is feasible for me to use `eventsource-parser` in other projects.

Reply:
Returning the same value will break OpenAI/other compatible SDKs as they expect a different structure. We will have to return undefined in this case so that event is skipped from being sent to the clients.